api/stacks/deployer.go

```diff
@@ -0,0 +1,46 @@
+package stacks
+
+import (
+	"sync"
+
+	portainer "github.com/portainer/portainer/api"
+)
+
+type StackDeployer interface {
+	DeploySwarmStack(stack *portainer.Stack, endpoint *portainer.Endpoint, registries []portainer.Registry, prune bool) error
+	DeployComposeStack(stack *portainer.Stack, endpoint *portainer.Endpoint, registries []portainer.Registry) error
+}
+
+type stackDeployer struct {
+	lock                *sync.Mutex
+	swarmStackManager   portainer.SwarmStackManager
+	composeStackManager portainer.ComposeStackManager
+}
+
+func NewStackDeployer(swarmStackManager portainer.SwarmStackManager, composeStackManager portainer.ComposeStackManager) *stackDeployer {
+	return &stackDeployer{
+		lock:                &sync.Mutex{},
+		swarmStackManager:   swarmStackManager,
+		composeStackManager: composeStackManager,
+	}
+}
+
+func (d *stackDeployer) DeploySwarmStack(stack *portainer.Stack, endpoint *portainer.Endpoint, registries []portainer.Registry, prune bool) error {
```

Comment:
why not a general method `DeployStack` that takes the same parameters and deploys the stack based on its type?

Reply:
Swarm has an extra param of `prune` that compose doesn't have, so params already aren't exactly the same. With the introduction of Kube, they may diverge even further.
It may be that we can find a nice way to bring all them together under the same umbrella, as for now, I've left them separated.

Reply:
as far as I have seen, kube doesn't add more. We replace the last parameter with `DeployOptions` that currently have only `prune`.

It would be nice to replace the all exec services to be internal services of this service.

but, we can do it later :)